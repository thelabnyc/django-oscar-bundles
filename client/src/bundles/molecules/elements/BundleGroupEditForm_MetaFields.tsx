import React from "react";
import classNames from "classnames";
import {
    IBundleGroup,
    IDRFSelectOptions,
} from "../../../utils/models.interfaces";

interface IProps {
    bundleTypeChoices: IDRFSelectOptions;
    group: IBundleGroup | null;
    isSaving: boolean;
    errors: {
        bundleType?: string[];
        isActive?: string[];
        name?: string[];
        headline?: string[];
        description?: string[];
        image?: string[];
    };
    bundleType: string;
    name: string;
    headline: string;
    isActive: boolean;
    description: string;
    image: File | null;
    clearImage: boolean;

    onEdit: (
        e: React.FormEvent<
            HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement
        >,
    ) => void;
    onSelectImage: (e: React.FormEvent<HTMLInputElement>) => void;
    onClearImage: (e: React.FormEvent<HTMLInputElement>) => void;
}

interface IState {}

export class BundleGroupMetaFields extends React.PureComponent<IProps, IState> {
    private buildFormGroupClasses(field: keyof IProps["errors"]) {
        const fieldErrors = this.props.errors[field];
        const classes: { [name: string]: boolean } = {
            "form-group": true,
            "has-error": fieldErrors ? fieldErrors.length > 0 : false,
        };
        return classNames(classes);
    }

    private buildErrors(field: keyof IProps["errors"]) {
        return (this.props.errors[field] || []).map((errorMsg, i) => {
            return (
                <span key={i} className="help-block">
                    <i className="fas fa-exclamation"></i> {errorMsg}
                </span>
            );
        });
    }

    private buildCurrentImage() {
        if (!this.props.group || !this.props.group.image) {
            return null;
        }
        const pathParts = this.props.group.image.split("/");
        const fileName = pathParts[pathParts.length - 1];
        return (
            <span>
                {gettext("Currently:")}{" "}
                <a
                    target="_blank"
                    rel="noreferrer"
                    href={this.props.group.image}
                >
                    {fileName}
                </a>{" "}
                <input
                    type="checkbox"
                    id="bundle-edit-form-clear-img"
                    checked={this.props.clearImage}
                    onChange={this.props.onClearImage}
                    disabled={this.props.isSaving}
                />{" "}
                <label htmlFor="bundle-edit-form-clear-img">
                    {gettext("Clear")}
                </label>
            </span>
        );
    }

    render() {
        const title = this.props.group
            ? gettext("Edit Bundle Group")
            : gettext("Create Bundle Group");
        return (
            <div className="col-sm-12 bundle-group-edit__section">
                <h1>{title}</h1>
                <div className={this.buildFormGroupClasses("bundleType")}>
                    <label htmlFor="id_bundleType" className="control-label">
                        {gettext("Bundle Type")}
                    </label>
                    <div>
                        <select
                            id="id_bundleType"
                            name="bundleType"
                            className="form-control"
                            value={this.props.bundleType}
                            onChange={this.props.onEdit}
                            disabled={this.props.isSaving}
                        >
                            {this.props.bundleTypeChoices.map((choice) => {
                                return (
                                    <option
                                        key={choice.value}
                                        value={choice.value}
                                    >
                                        {choice.display_name}
                                    </option>
                                );
                            })}
                        </select>
                        {this.buildErrors("bundleType")}
                    </div>
                </div>
                <div className={this.buildFormGroupClasses("name")}>
                    <label htmlFor="id_name" className="control-label">
                        {gettext("Name")}
                    </label>
                    <div>
                        <input
                            id="id_name"
                            name="name"
                            maxLength={200}
                            className="form-control"
                            value={this.props.name}
                            onChange={this.props.onEdit}
                            disabled={this.props.isSaving}
                        />
                        {this.buildErrors("name")}
                    </div>
                </div>
                <div className={this.buildFormGroupClasses("headline")}>
                    <label htmlFor="id_headline" className="control-label">
                        {gettext("Headline")}
                    </label>
                    <div>
                        <textarea
                            id="id_headline"
                            name="headline"
                            cols={40}
                            rows={4}
                            className="form-control"
                            value={this.props.headline}
                            onChange={this.props.onEdit}
                            disabled={this.props.isSaving}
                        ></textarea>
                        {this.buildErrors("headline")}
                    </div>
                </div>
                <div className={this.buildFormGroupClasses("isActive")}>
                    <label htmlFor="id_active_status" className="control-label">
                        {gettext("Active Status")}
                    </label>
                    <div>
                        <select
                            id="id_isActive"
                            name="isActive"
                            className="form-control"
                            value={String(this.props.isActive)}
                            onChange={this.props.onEdit}
                            disabled={this.props.isSaving}
                        >
                            {[
                                { display_name: "Active", value: "true" },
                                { display_name: "Inactive", value: "false" },
                            ].map((choice, index) => {
                                return (
                                    <option key={index} value={choice.value}>
                                        {choice.display_name}
                                    </option>
                                );
                            })}
                        </select>
                        {this.buildErrors("isActive")}
                    </div>
                </div>
                <div className={this.buildFormGroupClasses("description")}>
                    <label htmlFor="id_description" className="control-label">
                        {gettext("Description")}
                    </label>
                    <div>
                        <textarea
                            id="id_description"
                            name="description"
                            cols={40}
                            rows={4}
                            className="form-control"
                            value={this.props.description}
                            onChange={this.props.onEdit}
                            disabled={this.props.isSaving}
                        ></textarea>
                        {this.buildErrors("description")}
                    </div>
                </div>
                <div className={this.buildFormGroupClasses("image")}>
                    <label htmlFor="id_image" className="control-label">
                        {gettext("Image")}
                    </label>
                    <div>
                        {this.buildCurrentImage()}
                        <input
                            id="id_image"
                            type="file"
                            name="image"
                            className="form-control"
                            onChange={this.props.onSelectImage}
                            disabled={this.props.isSaving}
                        />
                        {this.buildErrors("image")}
                    </div>
                </div>
            </div>
        );
    }
}
