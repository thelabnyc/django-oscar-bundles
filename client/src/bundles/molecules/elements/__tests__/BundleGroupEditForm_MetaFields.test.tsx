import React from 'react';
import {shallow} from 'enzyme';
import {BundleGroupMetaFields} from '../BundleGroupEditForm_MetaFields';


describe('#bundles/molecules/elements/BundleGroupMetaFields', () => {

    it('renders', () => {
        const wrapper = shallow(
            <BundleGroupMetaFields
                bundleTypeChoices={[
                    {
                        display_name: 'Default',
                        value: 'default',
                    },
                ]}
                group={{
                    id: 1,
                    bundle_type: 'default',
                    name: 'name',
                    description: '',
                    headline: '',
                    image: 'https://placeimg.com/640/480/any',
                    newImage: null,
                    clearImage: false,
                    triggering_parents: [],
                    suggested_parents: [],
                    concrete_bundles: [],
                    user_configurable_bundles: [],
                }}
                isSaving={false}
                errors={{}}
                bundleType={''}
                name={''}
                headline={''}
                description={''}
                image={null}
                clearImage={false}
                onEdit={() => null}
                onSelectImage={() => null}
                onClearImage={() => null}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });


    it('renders error messages', () => {
        const wrapper = shallow(
            <BundleGroupMetaFields
                bundleTypeChoices={[]}
                group={null}
                isSaving={false}
                errors={{
                    bundleType: [],
                    name: ["Please choose a valid name"],
                    headline: [],
                    description: [],
                    image: [],
                }}
                bundleType={''}
                name={''}
                headline={''}
                description={''}
                image={null}
                clearImage={false}
                onEdit={() => null}
                onSelectImage={() => null}
                onClearImage={() => null}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

});
