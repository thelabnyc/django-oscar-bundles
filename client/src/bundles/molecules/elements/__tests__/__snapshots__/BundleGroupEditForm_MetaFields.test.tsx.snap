// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#bundles/molecules/elements/BundleGroupMetaFields renders 1`] = `
<div
  className="col-sm-12 bundle-group-edit__section"
>
  <h1>
    Edit Bundle Group
  </h1>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="id_bundleType"
    >
      Bundle Type
    </label>
    <div>
      <select
        className="form-control"
        disabled={false}
        id="id_bundleType"
        name="bundleType"
        onChange={[Function]}
        value=""
      >
        <option
          key="default"
          value="default"
        >
          Default
        </option>
      </select>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="id_name"
    >
      Name
    </label>
    <div>
      <input
        className="form-control"
        disabled={false}
        id="id_name"
        maxLength={200}
        name="name"
        onChange={[Function]}
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="id_headline"
    >
      Headline
    </label>
    <div>
      <textarea
        className="form-control"
        cols={40}
        disabled={false}
        id="id_headline"
        name="headline"
        onChange={[Function]}
        rows={4}
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="id_description"
    >
      Description
    </label>
    <div>
      <textarea
        className="form-control"
        cols={40}
        disabled={false}
        id="id_description"
        name="description"
        onChange={[Function]}
        rows={4}
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="id_image"
    >
      Image
    </label>
    <div>
      <span>
        Currently:
         
        <a
          href="https://placeimg.com/640/480/any"
          target="_blank"
        >
          any
        </a>
         
        <input
          checked={false}
          disabled={false}
          id="bundle-edit-form-clear-img"
          onChange={[Function]}
          type="checkbox"
        />
         
        <label
          htmlFor="bundle-edit-form-clear-img"
        >
          Clear
        </label>
      </span>
      <input
        className="form-control"
        disabled={false}
        id="id_image"
        name="image"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
</div>
`;

exports[`#bundles/molecules/elements/BundleGroupMetaFields renders error messages 1`] = `
<div
  className="col-sm-12 bundle-group-edit__section"
>
  <h1>
    Create Bundle Group
  </h1>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="id_bundleType"
    >
      Bundle Type
    </label>
    <div>
      <select
        className="form-control"
        disabled={false}
        id="id_bundleType"
        name="bundleType"
        onChange={[Function]}
        value=""
      />
    </div>
  </div>
  <div
    className="form-group has-error"
  >
    <label
      className="control-label"
      htmlFor="id_name"
    >
      Name
    </label>
    <div>
      <input
        className="form-control"
        disabled={false}
        id="id_name"
        maxLength={200}
        name="name"
        onChange={[Function]}
        value=""
      />
      <span
        className="help-block"
        key="0"
      >
        <i
          className="icon-exclamation-sign"
        />
         
        Please choose a valid name
      </span>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="id_headline"
    >
      Headline
    </label>
    <div>
      <textarea
        className="form-control"
        cols={40}
        disabled={false}
        id="id_headline"
        name="headline"
        onChange={[Function]}
        rows={4}
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="id_description"
    >
      Description
    </label>
    <div>
      <textarea
        className="form-control"
        cols={40}
        disabled={false}
        id="id_description"
        name="description"
        onChange={[Function]}
        rows={4}
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="id_image"
    >
      Image
    </label>
    <div>
      <input
        className="form-control"
        disabled={false}
        id="id_image"
        name="image"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
</div>
`;
